runtime/lua/vim/treesitter/highlighter.lua

```diff
@@ -258,51 +258,69 @@ local function on_line_impl(self, buf, line, is_spell_nav)
 
     while line >= state.next_row do
-      local capture, node, metadata = state.iter(line)
+      local pattern, match, metadata = state.iter()
 
-      local range = { root_end_row + 1, 0, root_end_row + 1, 0 }
-      if node then
-        range = vim.treesitter.get_range(node, buf, metadata and metadata[capture])
+      if not match then
+        state.next_row = root_end_row + 1
       end
-      local start_row, start_col, end_row, end_col = Range.unpack4(range)
 
-      if capture then
-        local hl = state.highlighter_query:get_hl_from_capture(capture)
-
-        local capture_name = state.highlighter_query:query().captures[capture]
-        local spell = nil ---@type boolean?
-        if capture_name == 'spell' then
-          spell = true
-        elseif capture_name == 'nospell' then
-          spell = false
-        end
-
-        -- Give nospell a higher priority so it always overrides spell captures.
-        local spell_pri_offset = capture_name == 'nospell' and 1 or 0
-
-        if hl and end_row >= line and (not is_spell_nav or spell ~= nil) then
-          local priority = (tonumber(metadata.priority) or vim.highlight.priorities.treesitter)
-            + spell_pri_offset
-          api.nvim_buf_set_extmark(buf, ns, start_row, start_col, {
-            end_line = end_row,
-            end_col = end_col,
-            hl_group = hl,
-            ephemeral = true,
-            priority = priority,
-            conceal = metadata.conceal,
-            spell = spell,
-          })
+      for capture, nodes in pairs(match or {}) do
+        for _, node in ipairs(nodes) do
+          local range = vim.treesitter.get_range(node, buf, metadata and metadata[capture])
+          local start_row, start_col, end_row, end_col = Range.unpack4(range)
+
+          local hl = state.highlighter_query:get_hl_from_capture(capture)
+
+          local capture_name = state.highlighter_query:query().captures[capture]
+          local spell = nil ---@type boolean?
+          if capture_name == 'spell' then
+            spell = true
+          elseif capture_name == 'nospell' then
+            spell = false
+          end
+
+          -- Give nospell a higher priority so it always overrides spell captures.
+          local spell_pri_offset = capture_name == 'nospell' and 1 or 0
+
+          if hl and end_row >= line and (not is_spell_nav or spell ~= nil) then
+            local priority = (tonumber(metadata.priority) or vim.highlight.priorities.treesitter)
+              + spell_pri_offset
+            local url = metadata[capture] and metadata[capture].url ---@type string|number|nil
+            if type(url) == 'number' then
+              if match and match[url] then
+                -- Assume there is only one matching node. If there are more than one, take the URL
+                -- from the first.
+                local other_node = match[url][1]
+                url = vim.treesitter.get_node_text(other_node, buf, metadata and metadata[url])
```

Comment:
Q: Is this line covered by tests? `get_node_text()` expects the third argument to be `opt = { metadata = ... }`, so it looks like we are not passing the correct argument.

Reply:
Alas, it was not that easy. I suspect the problem lies deeper (another bug in `iter_matches`).

Reply:
Ok I think I understand what's happening. The vimdoc query uses patterns like this:

```
(tag
  "*" @markup.heading.5.marker
  (#set! conceal "")
  text: (_) @label)
```

But with `iter_matches`, the `#set!` directive now applies to the *whole pattern*, whereas with `iter_captures` it would have applied only to the previous capture(?).

Reply:
Hmm, that's a fairly common pattern in nvim-treesitter (also with predicates).

Reply:
The solution is to make it more explicit about _what_ should be concealed:

```
(taglink
  "|" @markup.link.delimiter
  (#set! @markup.link.delimiter conceal "")
  text: (_) @markup.link)
```

This is (imo) more explicit and easier to understand. But at the obvious cost of being (yet another) breaking change.

Reply:
I consider that acceptable (since indeed better because more explicit). And fairly straightforward to fix, though tedious.

Reply:
For highlight queries; adapting injections would be a mess...

Reply:
(Probably also applies to priority.)

Reply:
Please never mind about Markdown. I think conceals are fine with Markdown. For injections I will try to find a repro and let you know in another place.

Using the new vimdoc highlight query, conceals in vimdoc now look fine. One concern is that this new query doesn't work fully backward-compatible with the previous highlighter implementation (nvim 0.9.x or on the current master): "`", `*`, `|` are no longer hidden. This might be a bit problematic because the query files from `nvim-treesitter` also need to be updated.

Reply:
I trust @clason to handle the situation with nvim-treesitter. When this PR is merged we can update the queries on the `main` branch of nvim-treesitter which is intended to work with 0.10 only.

Reply:
Might have to bite the bullet and backport (part of) this.